### PR TITLE
Backup restore

### PR DESCRIPTION
**Pull request by hobleyd**
Hey,

I noticed that backup and restore of settings didn't work reliably due to the DB needing to be closed prior to the relevant function executing. I have also tried to rehydrate the Widgets stored in the DB; I think I have this working. 

In the case where we can't bind the app widget id (see code) then it rehydrates the widget, but doesn't position it correctly. Not sure why.

On to merging the app drawer(s) next.

Cheers,
David

### CODE REVIEW

File: app/src/main/java/com/benny/openlauncher/viewutil/ItemViewFactory.java [hunk 1]
```diff
@@ -81,7 +89,35 @@ public static View getItemView(final Context context, final DesktopCallback call
...
+                    appWidgetInfo = HomeActivity._appWidgetManager.getAppWidgetInfo(appWidgetId);
+                    item.setWidgetValue(appWidgetId);
+                    HomeActivity._db.updateItem(item);
+                } else {
+                    LOG.error("Unable to bind app widget id: {} ", cn);
+                    Intent intent = new Intent(AppWidgetManager.ACTION_APPWIDGET_BIND);
+                    intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
+                    intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_PROVIDER, cn);
+
+                    HomeActivity._launcher.startActivityForResult(intent, HomeActivity.REQUEST_PICK_APPWIDGET);
```
Comment (dkanada) [inline app/src/main/java/com/benny/openlauncher/viewutil/ItemViewFactory.java]: I don't think we should open this dialog without user input, I would just reduce the nesting here and delete the item if the widget can't be rehydrated.

File: app/src/main/java/com/benny/openlauncher/viewutil/ItemViewFactory.java [hunk 2]
```diff
@@ -81,7 +89,35 @@ public static View getItemView(final Context context, final DesktopCallback call
...
+                    return null;
+                }
+            } else {
+                // Delete the Widget if we don't have enough information to rehydrate it.
+                HomeActivity._db.deleteItem(item, false);
```
Comment (dkanada) [inline app/src/main/java/com/benny/openlauncher/viewutil/ItemViewFactory.java]: please add a log statement here as well :)

File: app/src/main/java/com/benny/openlauncher/activity/SettingsActivity.java [hunk 3]
```diff
@@ -65,10 +66,12 @@ public boolean onPreferenceStartFragment(PreferenceFragmentCompat caller, Prefer
     protected void onActivityResult(int requestCode, int resultCode, final Intent data) {
         super.onActivityResult(requestCode, resultCode, data);
         if (resultCode == RESULT_OK) {
+            Setup.dataManager().close();
```
Comment (dkanada) [inline app/src/main/java/com/benny/openlauncher/activity/SettingsActivity.java]: You didn't create this method, is this in the parent class?

File: app/src/main/java/com/benny/openlauncher/model/Item.java [hunk 4]
```diff
@@ -79,9 +81,10 @@ public static Item newActionItem(int action) {
         return item;
     }
 
-    public static Item newWidgetItem(int widgetValue) {
+    public static Item newWidgetItem(ComponentName cn, int widgetValue) {
```
Comment (dkanada) [inline app/src/main/java/com/benny/openlauncher/model/Item.java]: maybe rename to componentName

File: app/src/main/java/com/benny/openlauncher/viewutil/ItemViewFactory.java (hunk 1 above)
Comment (hobleyd) [reply app/src/main/java/com/benny/openlauncher/viewutil/ItemViewFactory.java]: There appear to be 3 times that we need to rehydrate widgets:
1. We backup and restore in the running process. This just works and there is no user input.
2. We reinstall the application with the same signature; I have managed to trigger the bindAppWidgetIdIfAllowed() success path and I think that was how I did it. Although it's strange, the documentation says that this should only work if it is a System App. At least as I read it. I put it in for completeness, but did see a successful test through it.
3. We reinstall the application with a different signature; this helps me mainly honestly - I am testing on my primary device and use 2 machines for working on it (depending on whether I am on the road, or at home). So I find it quite useful. In terms of user input, I click on restore; go back from Settings and then the permissions dialog comes up, so I am not sure it is too disconcerting.

I can remove it if you are sure, but would be keen to keep it for a bit as I use it regularly. It isn't a big issue to manually put them back in again I guess though. Let me know.

File: app/src/main/java/com/benny/openlauncher/activity/SettingsActivity.java (hunk 3 above)
Comment (hobleyd) [reply app/src/main/java/com/benny/openlauncher/activity/SettingsActivity.java]: Yes it is.

File: app/src/main/java/com/benny/openlauncher/model/Item.java (hunk 4 above)
Comment (hobleyd) [reply app/src/main/java/com/benny/openlauncher/model/Item.java]: Done.

File: app/src/main/java/com/benny/openlauncher/viewutil/ItemViewFactory.java (hunk 2 above)
Comment (hobleyd) [reply app/src/main/java/com/benny/openlauncher/viewutil/ItemViewFactory.java]: Done.

File: app/src/main/java/com/benny/openlauncher/viewutil/ItemViewFactory.java (hunk 1 above)
Comment (dkanada) [reply app/src/main/java/com/benny/openlauncher/viewutil/ItemViewFactory.java]: We can leave it there for now (I don't use widgets so this doesn't affect me :p) and if anyone complains I'll think about removing it then.

**Pull request merged**